File: app/lib/services/goal_service.dart
import '../database/database_helper.dart';
import '../utils/constants.dart';
import 'order_service.dart';

class GoalProgress {
  final double goal;
  final double current;
  final double remaining;
  final int orderCount;
  final int? ordersNeeded;
  final int percent;

  const GoalProgress({
    required this.goal,
    required this.current,
    required this.remaining,
    required this.orderCount,
    this.ordersNeeded,
    required this.percent,
  });
}

class GoalService {
  final DatabaseHelper _db = DatabaseHelper();
  final OrderService _orders = OrderService();

  Future<double?> getMonthlyGoal() async {
    final v = await _db.getSetting(AppConstants.keyMonthlyRevenueGoal);
    if (v == null || v.isEmpty) return null;
    final n = double.tryParse(v);
    return (n != null && n > 0) ? n : null;
  }

  Future<void> setMonthlyGoal(double amount) async {
    if (amount <= 0) {
      await _db.setSetting(AppConstants.keyMonthlyRevenueGoal, '');
      return;
    }
    await _db.setSetting(
      AppConstants.keyMonthlyRevenueGoal,
      amount.round().toString(),
    );
  }

  Future<GoalProgress?> getGoalProgress([DateTime? now]) async {
    final goal = await getMonthlyGoal();
    if (goal == null) return null;

    final d = now ?? DateTime.now();
    final summary = await _orders.getSalesSummaryForMonth(d);
    final current = summary.revenue;
    final remaining = (goal - current).clamp(0.0, double.infinity).toDouble();
    final avg = summary.orderCount > 0 ? current / summary.orderCount : 0.0;
    final int? ordersNeeded = remaining > 0 && avg > 0
        ? (remaining / avg).ceil()
        : remaining > 0
            ? null
            : 0;

    return GoalProgress(
      goal: goal,
      current: current,
      remaining: remaining,
      orderCount: summary.orderCount,
      ordersNeeded: ordersNeeded,
      percent: (current / goal * 100).round().clamp(0, 100),
    );
  }
}

File: web/src/lib/goalService.ts
import { getSetting, setSetting } from './db';
import { SETTING_KEYS } from './constants';
import { getSalesSummaryForMonth } from './orderService';

export interface GoalProgress {
  goal: number;
  current: number;
  remaining: number;
  orderCount: number;
  ordersNeeded: number | null;
  percent: number;
}

export async function getMonthlyGoal(): Promise<number | null> {
  const v = await getSetting(SETTING_KEYS.monthlyRevenueGoal);
  if (!v) return null;
  const n = parseInt(v, 10);
  return n > 0 ? n : null;
}

export async function setMonthlyGoal(amount: number): Promise<void> {
  if (amount <= 0) {
    await setSetting(SETTING_KEYS.monthlyRevenueGoal, '');
    return;
  }
  await setSetting(SETTING_KEYS.monthlyRevenueGoal, String(amount));
}

export async function getGoalProgress(now = new Date()): Promise<GoalProgress | null> {
  const goal = await getMonthlyGoal();
  if (!goal) return null;

  const summary = await getSalesSummaryForMonth(now);
  const current = summary.revenue;
  const remaining = Math.max(0, goal - current);
  const avg = summary.order_count > 0 ? current / summary.order_count : 0;
  const ordersNeeded = remaining > 0 && avg > 0 ? Math.ceil(remaining / avg) : remaining > 0 ? null : 0;

  return {
    goal,
    current,
    remaining,
    orderCount: summary.order_count,
    ordersNeeded,
    percent: Math.min(100, Math.round((current / goal) * 100)),
  };
}
